Web Inspector: Debugger: symbolic breakpoints should work with intrinsic functions

dcrousso opened the pull request:
#### 06ff018c22a0dfacae6584f2cc7d51568947ff6c
<pre>
Web Inspector: Debugger: symbolic breakpoints should work with intrinsic functions
<a href="https://bugs.webkit.org/show_bug.cgi?id=243994">https://bugs.webkit.org/show_bug.cgi?id=243994</a>
&lt;<a href="https://rdar.apple.com/problem/99037335">rdar://problem/99037335</a>&gt;

Reviewed by Patrick Angle and Yusuke Suzuki.

Symbolic Breakpoints already work with `NativeExecutable` by swapping the JIT thunk with an equivalent version that talks to the `Debugger`.

Specifically, there&apos;s a code pointer each on the `NativeJITCode` and `NativeExecutable`.

Intrinsics, however, have an additional code pointer that must be swapped in `DirectJITCode`.

* Source/JavaScriptCore/jit/JITCode.h:
(JSC::JITCode::canSwapCodePtrWithArityCheckForDebugger const): Added.
* Source/JavaScriptCore/jit/JITCode.cpp:
(JSC::JITCode::swapCodePtrWithArityCheckForDebugger): Added.
(JSC::JITCodeWithCodeRef::swapCodeRefForDebugger):
(JSC::DirectJITCode::swapCodePtrWithArityCheckForDebugger): Added.
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::ReplacedThunk::~ReplacedThunk):
(Inspector::InspectorDebuggerAgent::didCreateNativeExecutable):

* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-insensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-insensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-sensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-sensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-insensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-insensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-sensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-sensitive-expected.txt: Added.

* LayoutTests/inspector/debugger/symbolic-breakpoint-native-js-exact-case-insensitive.html:

Drive-by: Fix incorrect symbol name.
Canonical link: <a href="https://commits.webkit.org/315713@main">https://commits.webkit.org/315713@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d0f5e47d3861cc97ac922607acfbd89d123a34f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/169669 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/43591 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/168/builds/36950 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/179028 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/127381 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/44655 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/43220 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/132217 "Passed tests") | [❌ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/93131 "2 flakes 10 failures") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/172628 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/35378 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/152593 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/112720 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/154/builds/34092 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/32355 "Passed tests") | [✅ 🛠 gtk3-libwebrtc](https://ews-build.webkit.org/#/builders/173/builds/27725 "Built successfully") | 
| [✅ 🧪 jsc-x86-64](https://ews-build.webkit.org/#/builders/218/builds/1010 "Passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/144453 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/169/builds/31992 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/181627 "Built successfully") | 
| [✅ 🛠 🧪 jsc-debug-arm64](https://ews-build.webkit.org/#/builders/171/builds/30924 "Built successfully and passed tests") | [✅ 🛠 ios-safer-cpp](https://ews-build.webkit.org/#/builders/174/builds/34335 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/36521 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/140664 "Passed tests") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/43247 "Built successfully") | [❌ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/152063 "") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/140500 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/37872 "Built successfully and passed tests") | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/43936 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/170/builds/28972 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/108176 "Built successfully") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/35767 "Passed tests") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/115701 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/202348 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/42446 "Built successfully") | [✅ 🧪 mac-site-isolation](https://ews-build.webkit.org/#/builders/185/builds/7065 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/54253 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/41839 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/42094 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/41982 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

dcrousso committed:
Web Inspector: Debugger: symbolic breakpoints should work with intrinsic functions
https://bugs.webkit.org/show_bug.cgi?id=243994
<rdar://problem/99037335>

Reviewed by Yusuke Suzuki.

This should bring symbolic breakpoint "coverage" up even higher.

* Source/JavaScriptCore/jit/ThunkGenerator.h:
* Source/JavaScriptCore/jit/ThunkGenerators.h:
* Source/JavaScriptCore/jit/ThunkGenerators.cpp:
(JSC::handleExceptionGenerator):
(JSC::handleExceptionWithCallFrameRollbackGenerator):
(JSC::popThunkStackPreservesAndHandleExceptionGenerator):
(JSC::checkExceptionGenerator):
(JSC::throwExceptionFromCallSlowPathGenerator):
(JSC::linkCallThunkGenerator):
(JSC::linkPolymorphicCallThunkGenerator):
(JSC::virtualThunkFor):
(JSC::virtualThunkForRegularCall):
(JSC::virtualThunkForTailCall):
(JSC::virtualThunkForConstructConstruct):
(JSC::nativeForGenerator):
(JSC::nativeCallGenerator):
(JSC::nativeTailCallGenerator):
(JSC::nativeTailCallWithoutSavedTagsGenerator):
(JSC::nativeConstructGenerator):
(JSC::internalFunctionCallGenerator):
(JSC::internalFunctionConstructGenerator):
(JSC::arityFixupGenerator):
(JSC::unreachableGenerator):
(JSC::stringGetByValGenerator):
(JSC::charCodeAtThunkGenerator):
(JSC::charAtThunkGenerator):
(JSC::fromCharCodeThunkGenerator):
(JSC::stringPrototypeCodePointAtThunkGenerator):
(JSC::clz32ThunkGenerator):
(JSC::sqrtThunkGenerator):
(JSC::floorThunkGenerator):
(JSC::ceilThunkGenerator):
(JSC::truncThunkGenerator):
(JSC::roundThunkGenerator):
(JSC::numberConstructorCallThunkGenerator):
(JSC::stringConstructorCallThunkGenerator):
(JSC::expThunkGenerator):
(JSC::logThunkGenerator):
(JSC::absThunkGenerator):
(JSC::imulThunkGenerator):
(JSC::randomThunkGenerator):
(JSC::boundFunctionCallGenerator):
(JSC::remoteFunctionCallGenerator):
(JSC::nativeCallWithDebuggerHookGenerator): Deleted.
(JSC::nativeConstructWithDebuggerHookGenerator): Deleted.
* Source/JavaScriptCore/jit/JITThunks.h:
* Source/JavaScriptCore/jit/JITThunks.cpp:
(JSC::JITThunks::ctiNativeCall):
(JSC::JITThunks::ctiNativeConstruct):
(JSC::JITThunks::ctiNativeTailCall):
(JSC::JITThunks::ctiNativeTailCallWithoutSavedTags):
(JSC::JITThunks::ctiInternalFunctionCall):
(JSC::JITThunks::ctiInternalFunctionConstruct):
(JSC::JITThunks::ctiStubImpl):
(JSC::JITThunks::ctiStub):
(JSC::JITThunks::ctiSlowPathFunctionStub):
(JSC::JITThunks::hostFunctionStub):
(JSC::JITThunks::ctiNativeCallWithDebuggerHook): Deleted.
(JSC::JITThunks::ctiNativeConstructWithDebuggerHook): Deleted.
* Source/JavaScriptCore/dfg/DFGThunks.h:
* Source/JavaScriptCore/dfg/DFGThunks.cpp:
(JSC::DFG::osrExitGenerationThunkGenerator):
(JSC::DFG::osrEntryThunkGenerator):
* Source/JavaScriptCore/ftl/FTLThunks.h:
* Source/JavaScriptCore/ftl/FTLThunks.cpp:
(JSC::FTL::osrExitGenerationThunkGenerator):
(JSC::FTL::lazySlowPathGenerationThunkGenerator):
* Source/JavaScriptCore/jit/JIT.h:
* Source/JavaScriptCore/jit/JIT.cpp:
(JSC::JIT::consistencyCheckGenerator):
* Source/JavaScriptCore/jit/JITCall.cpp:
(JSC::JIT::returnFromBaselineGenerator):
* Source/JavaScriptCore/jit/JITOpcodes.cpp:
(JSC::JIT::valueIsFalseyGenerator):
(JSC::JIT::valueIsTruthyGenerator):
(JSC::JIT::op_throw_handlerGenerator):
(JSC::JIT::op_enter_handlerGenerator):
(JSC::JIT::op_check_traps_handlerGenerator):
* Source/JavaScriptCore/jit/JITPropertyAccess.cpp:
(JSC::JIT::slow_op_get_by_val_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::slow_op_get_private_name_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::slow_op_put_by_val_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::slow_op_put_private_name_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::slow_op_del_by_id_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::slow_op_del_by_val_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::slow_op_get_by_id_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::slow_op_get_by_id_with_this_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::slow_op_put_by_id_callSlowOperationThenCheckExceptionGenerator):
(JSC::JIT::generateOpResolveScopeThunk):
(JSC::JIT::slow_op_resolve_scopeGenerator):
(JSC::JIT::generateOpGetFromScopeThunk):
(JSC::JIT::slow_op_get_from_scopeGenerator):
(JSC::JIT::slow_op_put_to_scopeGenerator):
(JSC::JIT::slow_op_get_by_val_with_this_callSlowOperationThenCheckExceptionGenerator):
* Source/JavaScriptCore/runtime/VM.h:
* Source/JavaScriptCore/runtime/VM.cpp:
(JSC::VM::thunkGeneratorForIntrinsic):
(JSC::VM::getCTIStub):
(JSC::VM::getHostFunction):
(JSC::VM::getCTIInternalFunctionTrampolineFor):
Instead of having entirely separate ways to generate JIT thunks with debugger hooks (e.g. having a
`ctiNativeCallWithDebuggerHook` alongside a `ctiNativeCall`), require that all `ThunkGenerator` have
a `IncludeDebuggerHook` as a second paramenter, including that in the hash when caching the result.
This way, there only one path to generating JIT thunks (and future engineers will be reminded to
consider the debugging story (if any) instead of potentially leaving that aside). Only JIT thunks
generated by generators defined in `VM::thunkGeneratorForIntrinsic` currently need to use this new
`IncludeDebuggerHook` parameter, as otherwise they either aren't a `NativeExecutable` entry point or
go down the already supported path for `NativeJITCode`/`NativeDOMJITCode`.

* Source/JavaScriptCore/jit/CCallHelpers.h:
* Source/JavaScriptCore/jit/CCallHelpers.cpp:
(JSC::CCallHelpers::emitDebuggerHook): Added.
Add a utility function to emit JIT code that calls into `operationDebuggerWillCallNativeExecutable`.

* Source/JavaScriptCore/jit/SpecializedThunkJIT.h:
(JSC::SpecializedThunkJIT::checkArgCount): Added.
Separate the argument count check into it's own callable utility function so that the existing
callsites can wait to invoke it until after any debugger hooks are added.

* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::functionName):
(Inspector::ReplacedThunk::~ReplacedThunk):
(Inspector::InspectorDebuggerAgent::addSymbolicBreakpoint):
(Inspector::InspectorDebuggerAgent::didCreateNativeExecutable):
* Source/JavaScriptCore/jit/JITCode.h:
(JSC::JITCode::canSwapCodePtrWithArityCheckForDebugger const): Added.
* Source/JavaScriptCore/jit/JITCode.cpp:
(JSC::JITCode::swapCodePtrWithArityCheckForDebugger): Added.
(JSC::JITCodeWithCodeRef::swapCodeRefForDebugger):
(JSC::DirectJITCode::swapCodePtrWithArityCheckForDebugger): Added.
Add a way for the `Debugger` (really the `InspectorDebuggerAgent`) to replace the `m_withArityCheck`
of `DirectJITCode` with a version generated with `IncludeDebuggerHook::Yes`.

* Source/JavaScriptCore/jit/JITOperations.h
* Source/JavaScriptCore/jit/JITOperations.cpp:
(JSC::operationDebuggerWillCallNativeExecutable):
Switch to getting the `CallFrame` from the `VM` instead of passing it in directly.

* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-insensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-insensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-sensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-sensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-insensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-insensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-sensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-sensitive-expected.txt: Added.

* LayoutTests/inspector/debugger/symbolic-breakpoint-native-js-exact-case-insensitive.html:
Drive-by: Fix incorrect symbol name.

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7bc72e75cb2f5079f34dd06040ee3dd3509a1070. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/5877

dcrousso commented:
AFAICT the `CallFrame` is valid based on logging inside `operationDebuggerWillCallNativeExecutable`, but I am seeing crashes like

```
Exception Codes:  KERN_INVALID_ADDRESS at 0xd29b4c88d503237f -> 0xffffcc88d503237f (possible pointer authentication failure)

0   0x136be92d8 JSC::StructureID::decontaminate() const + 24 (StructureID.h:79)
1   0x136be91f8 JSC::StructureID::decode() const + 40 (StructureID.h:130)
2   0x136be91f8 JSC::StructureID::decode() const + 40 (StructureID.h:130)
3   0x137aa2a1c JSC::JSCell::structure() const + 28 (JSCellInlines.h:137)
4   0x1382162b0 JSC::slowValidateCell(JSC::JSCell*) + 128 (JSCell.cpp:188)
5   0x136bef03c void JSC::validateCell<JSC::JSCell*>(JSC::JSCell*) + 28 (WriteBarrier.h:64)
6   0x1377790bc JSC::WriteBarrierBase<JSC::ScriptExecutable, WTF::RawPtrTraits<JSC::ScriptExecutable> >::get() const + 60 (WriteBarrier.h:103)
7   0x13792d4c0 JSC::CodeBlock::ownerExecutable() const + 32 (CodeBlock.h:364)
8   0x13742d104 JSC::DebuggerCallFrame::sourceIDForCallFrame(JSC::CallFrame*) + 484 (DebuggerCallFrame.cpp:326)
9   0x13742c548 JSC::Debugger::pauseIfNeeded(JSC::JSGlobalObject*) + 948 (Debugger.cpp:958)
10  0x13742c15c JSC::Debugger::breakProgram(WTF::RefPtr<JSC::Breakpoint, WTF::RawPtrTraits<JSC::Breakpoint>, WTF::DefaultRefDerefTraits<JSC::Breakpoint> >&&) + 1028 (Debugger.cpp:852)
11  0x137bb89d0 Inspector::InspectorDebuggerAgent::breakProgram(Inspector::DebuggerFrontendDispatcher::Reason, WTF::RefPtr<WTF::JSONImpl::Object, WTF::RawPtrTraits<WTF::JSONImpl::Object>, WTF::DefaultRefDerefTraits<WTF::JSONImpl::Object> >&&, WTF::RefPtr<JSC::Breakpoint, WTF::RawPtrTraits<JSC::Breakpoint>, WTF::DefaultRefDerefTraits<JSC::Breakpoint> >&&) + 104 (InspectorDebuggerAgent.cpp:1812)
12  0x137bc03b4 Inspector::InspectorDebuggerAgent::willCallNativeExecutable(JSC::CallFrame*) + 804 (InspectorDebuggerAgent.cpp:1542)
13  0x137441354 JSC::Debugger::willCallNativeExecutable(JSC::CallFrame*)::$_3::operator()(JSC::Debugger::Observer&) const + 80 (Debugger.cpp:266)
14  0x1374412a0 WTF::Detail::CallableWrapper<JSC::Debugger::willCallNativeExecutable(JSC::CallFrame*)::$_3, void, JSC::Debugger::Observer&>::call(JSC::Debugger::Observer&) + 56 (Function.h:53)
15  0x137427910 WTF::Function<void (JSC::Debugger::Observer&)>::operator()(JSC::Debugger::Observer&) const + 196 (Function.h:82)
16  0x137427370 JSC::Debugger::dispatchFunctionToObservers(WTF::Function<void (JSC::Debugger::Observer&)>) + 164 (Debugger.cpp:305)
17  0x13742743c JSC::Debugger::willCallNativeExecutable(JSC::CallFrame*) + 68 (Debugger.cpp:265)
18  0x137d16084 operationDebuggerWillCallNativeExecutable + 1128 (JITOperations.cpp:3683)
19  0x14601074c ???
20  0x146008248 ???
21  0x146008248 ???
22  0x146008848 ???
23  0x1384e9598 JSC::JITCode::execute(JSC::VM*, JSC::ProtoCallFrame*) + 216 (JITCodeInlines.h:42)
24  0x137c11864 JSC::Interpreter::executeProgram(JSC::SourceCode const&, JSC::JSGlobalObject*, JSC::JSObject*) + 7976 (Interpreter.cpp:1003)
25  0x1380bd3f8 JSC::evaluate(JSC::JSGlobalObject*, JSC::SourceCode const&, JSC::JSValue, WTF::NakedPtr<JSC::Exception>&) + 524 (Completion.cpp:138)
26  0x1380bd6a4 JSC::evaluateWithScopeExtension(JSC::JSGlobalObject*, JSC::SourceCode const&, JSC::JSObject*, WTF::NakedPtr<JSC::Exception>&) + 164 (Completion.cpp:165)
27  0x137b79438 Inspector::JSInjectedScriptHost::evaluateWithScopeExtension(JSC::JSGlobalObject*, JSC::CallFrame*) + 632 (JSInjectedScriptHost.cpp:127)
28  0x137b817c4 Inspector::jsInjectedScriptHostPrototypeFunctionEvaluateWithScopeExtension(JSC::JSGlobalObject*, JSC::CallFrame*) + 188 (JSInjectedScriptHostPrototype.cpp:262)
29  0x146018044 ???
30  0x146008008 ???
31  0x146008008 ???
32  0x146008848 ???
33  0x1384e9598 JSC::JITCode::execute(JSC::VM*, JSC::ProtoCallFrame*) + 216 (JITCodeInlines.h:42)
34  0x137c12710 JSC::Interpreter::executeCall(JSC::JSGlobalObject*, JSC::JSObject*, JSC::CallData const&, JSC::JSValue, JSC::ArgList const&) + 1552 (Interpreter.cpp:1072)
35  0x13802c194 JSC::call(JSC::JSGlobalObject*, JSC::JSValue, JSC::CallData const&, JSC::JSValue, JSC::ArgList const&) + 244 (CallData.cpp:57)
36  0x13802c250 JSC::call(JSC::JSGlobalObject*, JSC::JSValue, JSC::CallData const&, JSC::JSValue, JSC::ArgList const&, WTF::NakedPtr<JSC::Exception>&) + 168 (CallData.cpp:64)
37  0x282e886bc WebCore::JSExecState::call(JSC::JSGlobalObject*, JSC::JSValue, JSC::CallData const&, JSC::JSValue, JSC::ArgList const&, WTF::NakedPtr<JSC::Exception>&) + 184 (JSExecState.h:59)
38  0x282f24f80 WebCore::functionCallHandlerFromAnyThread(JSC::JSGlobalObject*, JSC::JSValue, JSC::CallData const&, JSC::JSValue, JSC::ArgList const&, WTF::NakedPtr<JSC::Exception>&) + 84 (JSExecState.cpp:46)
39  0x13710b168 Deprecated::ScriptFunctionCall::call() + 484 (ScriptFunctionCall.cpp:134)
40  0x137a98550 Inspector::InjectedScriptBase::callFunctionWithEvalEnabled(Deprecated::ScriptFunctionCall&) const + 84 (InjectedScriptBase.cpp:74)
41  0x137a96a9c Inspector::InjectedScriptBase::makeCall(Deprecated::ScriptFunctionCall&) + 124 (InjectedScriptBase.cpp:84)
42  0x137a95eb8 Inspector::InjectedScriptBase::makeEvalCall(WTF::String&, Deprecated::ScriptFunctionCall&, WTF::RefPtr<Inspector::Protocol::Runtime::RemoteObject, WTF::RawPtrTraits<Inspector::Protocol::Runtime::RemoteObject>, WTF::DefaultRefDerefTraits<Inspector::Protocol::Runtime::RemoteObject> >&, std::__1::optional<bool>&, std::__1::optional<int>&) + 72 (InjectedScriptBase.cpp:105)
43  0x137a96088 Inspector::InjectedScript::evaluate(WTF::String&, WTF::String const&, WTF::String const&, bool, bool, bool, bool, WTF::RefPtr<Inspector::Protocol::Runtime::RemoteObject, WTF::RawPtrTraits<Inspector::Protocol::Runtime::RemoteObject>, WTF::DefaultRefDerefTraits<Inspector::Protocol::Runtime::RemoteObject> >&, std::__1::optional<bool>&, std::__1::optional<int>&) + 344 (InjectedScript.cpp:81)
44  0x137bc5cc4 Inspector::InspectorRuntimeAgent::evaluate(Inspector::InjectedScript&, WTF::String const&, WTF::String const&, std::__1::optional<bool>&&, std::__1::optional<bool>&&, std::__1::optional<bool>&&, std::__1::optional<bool>&&, std::__1::optional<bool>&&, std::__1::optional<bool>&&) + 512 (InspectorRuntimeAgent.cpp:153)
45  0x28414ce20 WebCore::PageRuntimeAgent::evaluate(WTF::String const&, WTF::String const&, std::__1::optional<bool>&&, std::__1::optional<bool>&&, std::__1::optional<int>&&, std::__1::optional<bool>&&, std::__1::optional<bool>&&, std::__1::optional<bool>&&, std::__1::optional<bool>&&) + 440 (PageRuntimeAgent.cpp:200)
```

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a628f710c68a30cac85571517fa0817ccfb0ff17. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/5877

dcrousso commented:
after rebasing this and trying to run the new tests 1000 times (both individually and in sequence) I am not able to reproduce the above crash

justinmichaud commented:
From the commit message: 

>>> Reviewed by Yusuke Suzuki.

seems to be incorrect, unless I am missing something from a previous version of this patch

dcrousso commented:
> From the commit message:
> 
> > > > Reviewed by Yusuke Suzuki.
> 
> seems to be incorrect, unless I am missing something from a previous version of this patch

yeah the [original version of this PR that was rolled out](https://commits.webkit.org/254576@main) was reviewed by @Constellation, and this PR is pretty minimally different so I felt it was fine to keep it there

i can remove it if you'd prefer

patrickangle reviewed (approved):
This seems fine from a Web Inspector perspective - but please also get an r+ from @Constellation or someone else that works more closely on JSC before relanding

dcrousso committed:
Web Inspector: Debugger: symbolic breakpoints should work with intrinsic functions
https://bugs.webkit.org/show_bug.cgi?id=243994
<rdar://problem/99037335>

Reviewed by Patrick Angle and Yusuke Suzuki.

Symbolic Breakpoints already work with `NativeExecutable` by swapping the JIT thunk with an equivalent version that talks to the `Debugger`.

Specifically, there's a code pointer each on the `NativeJITCode` and `NativeExecutable`.

Intrinsics, however, have an additional code pointer that must be swapped in `DirectJITCode`.

* Source/JavaScriptCore/jit/JITCode.h:
(JSC::JITCode::canSwapCodePtrWithArityCheckForDebugger const): Added.
* Source/JavaScriptCore/jit/JITCode.cpp:
(JSC::JITCode::swapCodePtrWithArityCheckForDebugger): Added.
(JSC::JITCodeWithCodeRef::swapCodeRefForDebugger):
(JSC::DirectJITCode::swapCodePtrWithArityCheckForDebugger): Added.
* Source/JavaScriptCore/inspector/agents/InspectorDebuggerAgent.cpp:
(Inspector::ReplacedThunk::~ReplacedThunk):
(Inspector::InspectorDebuggerAgent::didCreateNativeExecutable):

* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-insensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-insensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-sensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-exact-case-sensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-insensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-insensitive-expected.txt: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-sensitive.html: Added.
* LayoutTests/inspector/debugger/symbolic-breakpoint-intrinsic-js-regex-case-sensitive-expected.txt: Added.

* LayoutTests/inspector/debugger/symbolic-breakpoint-native-js-exact-case-insensitive.html:

Drive-by: Fix incorrect symbol name.
Canonical link: https://commits.webkit.org/315713@main
